python/langsmith/client.py

```diff
@@ -3036,3 +3097,52 @@ def _evaluate_strings(self, prediction, reference=None, input=None, **kwargs) ->
             input_mapper=input_mapper,
             revision_id=revision_id,
         )
+
+
+def _tracing_thread_drain_queue(
+    tracing_queue: Queue, limit: Optional[int] = None
+) -> List[Tuple[str, dict]]:
+    next_batch: List[Tuple[str, dict]] = []
+    try:
+        while item := tracing_queue.get(block=True, timeout=0.25):
+            next_batch.append(item)
+            if limit and len(next_batch) >= limit:
+                break
+    except Empty:
+        pass
+    return next_batch
+
+
+def _tracing_thread_handle_batch(
+    client: Client, tracing_queue: Queue, batch: List[Tuple[str, dict]]
+) -> None:
+    create = [item[1] for item in batch if item[0] == "create"]
+    update = [item[1] for item in batch if item[0] == "update"]
+    try:
+        client.batch_ingest_runs(create=create, update=update, pre_sampled=True)
+    finally:
+        for _ in batch:
+            tracing_queue.task_done()
+
+
+def _tracing_thread_func(client_ref: weakref.ref[Client]) -> None:
+    client = client_ref()
+    if client is None:
+        return
+    tracing_queue = client.tracing_queue
+    assert tracing_queue is not None
+
+    # loop until
+    while (
+        # the main thread dies
+        threading.main_thread().is_alive()
```

Comment:
If we create the client in a separate thread, I guess this could keep running even after the calling thread dies?
We shouldn't be doing this I'm just making sure I understand

Reply:
No that's what the ref count is for. This main thread check is to stand in for a signal trap, if the main thread has died then we should die to let the process die